x/wasm/internal/keeper/genesis_test.go

```diff
@@ -67,6 +68,166 @@ func TestGenesisExportImport(t *testing.T) {
 	require.False(t, dstIT.Valid())
 }
 
+func TestFailFastImport(t *testing.T) {
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+	codeHash := sha256.Sum256(wasmCode)
+	anyAddress := make([]byte, 20)
+
+	specs := map[string]struct {
+		src        types.GenesisState
+		expSuccess bool
+	}{
+		"happy path: code info correct": {
+			src: types.GenesisState{
+				Codes: []types.Code{{
+					CodeInfo: wasmTypes.CodeInfo{
+						CodeHash: codeHash[:],
+						Creator:  anyAddress,
+					},
+					CodesBytes: wasmCode,
+				}},
+				Contracts: nil,
+			},
+			expSuccess: true,
+		},
+		"prevent code hash mismatch": {src: types.GenesisState{
+			Codes: []types.Code{{
+				CodeInfo: wasmTypes.CodeInfo{
+					CodeHash: make([]byte, len(codeHash)),
+					Creator:  anyAddress,
+				},
+				CodesBytes: wasmCode,
+			}},
+			Contracts: nil,
+		}},
+		"happy path: code info and contract do match": {
+			src: types.GenesisState{
+				Codes: []types.Code{{
+					CodeInfo: wasmTypes.CodeInfo{
+						CodeHash: codeHash[:],
```

Comment:
One issue still here, is the CodeID is not explicit in import or export. We rely on the ordering of the array.

So, if one CodeInfo was deleted, all references are broken. It would be nice to make this explicit (using a `GenesisCodeInfo` type or whatever), and use the explicit ID not the index in the array

Reply:
Ja, we should address this issue proper. I am afraid this is more work and should be priotized proper. I had planned a separate PR for it after some discussion. See "not covered" section.

Reply:
Okay, happy for a new issue on that. Can you make a proper issue for the "not covered" part and link it to this PR?

Reply:
Done #166